invoicer_project/user/views.py

```diff
@@ -65,6 +68,57 @@ def create(self, request):
             refresh = RefreshToken.for_user(user)
             return JsonResponse({'refresh': str(refresh), 'access': str(refresh.access_token)}, status=200)
         return JsonResponse({"message": "Invalid credentials"}, status=400)
+    
+    @action(detail=True, methods=['post'])
+    def send_confirmation_code(self, request):
+        data = JSONParser().parse(request)
+        email = data.get('email')
+        try:
+            user = User.objects.get(email=email)
+        except User.DoesNotExist:
+            user = None
+        if user:
+            code = ''
+            while not (any(c.isalpha() for c in code) and any(c.isdigit() for c in code)):
+                code = ''.join(random.choices(string.ascii_uppercase + string.digits, k=6))
+            subject = 'Confirmation Code'
+            message = f'Your confirmation code is {code}. It is valid for 1 minute.'
+            from_email = settings.EMAIL_HOST_USER
+            recipient_list = [email]
+            send_mail(subject, message, from_email, recipient_list)
+            cache.set(f'confirmation_code_{email}', code, timeout=60)
+            return JsonResponse({'message': 'Confirmation code sent'}, status=200)
+        else:
+            return JsonResponse({'error': 'Email address not found'}, status=404)
+        
+    def confirm_code(self, request):
+        data = JSONParser().parse(request)
+        email = data.get('email')
+        code = data.get('code')
+        cached_code = cache.get(f'confirmation_code_{email}')
+        if cached_code and cached_code == code:
+            cache.delete(f'confirmation_code_{email}')
+            return JsonResponse({'message': 'Confirmation successful'}, status=200)
+        else:
+            return JsonResponse({'message': 'Confirmation fail'}, status=404)
+        
+    def partial_update(self, request):
+        data = JSONParser().parse(request)
+        try:
+            user = User.objects.get(email=data['email'])
+        except User.DoesNotExist:
+            user = None
+        data['id'] = user.id
+        serializer = UserSettingsSerializer(user, data=data, partial=True)
+        if serializer.is_valid():
```

Comment:
It is also possible to invert the conditional here:
```
if not serializer.is_valid():
    return JsonResponse({'errors': serializer.errors}, status=400)
new_password = ...
```

Reply:
Done